<html><head><title>Repeated soft clustering for detection of empty clusters</title>
<link rel="stylesheet" type="text/css" href="Rchm.css">
</head>
<body>

<table width="100%"><tr><td>cselection(Mfuzz)</td><td align="right">R Documentation</td></tr></table><object type="application/x-oleobject" classid="clsid:1e2a7bd0-dab9-11d0-b93a-00c04fc99f9e">
<param name="keyword" value="R:   cselection">
<param name="keyword" value=" Repeated soft clustering for detection of empty clusters">
</object>


<h2>Repeated soft clustering for detection of empty clusters</h2>


<h3>Description</h3>

<p>
This function repeatly performs soft clustering
for a range of cluster numbers and reports the number of empty clusters
detected.
</p>


<h3>Usage</h3>

<pre>cselection(eset,m,crange=seq(4,32,4),repeats=5,visu=TRUE,...)</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>eset</code></td>
<td>
object of class &ldquo;exprSet&rdquo;</td></tr>
<tr valign="top"><td><code>m</code></td>
<td>
value of fuzzy c-means parameter <code>m</code></td></tr>
<tr valign="top"><td><code>crange</code></td>
<td>
range of number of clusters <code>c</code></td></tr>
<tr valign="top"><td><code>repeats</code></td>
<td>
number of repeated clusterings</td></tr>
<tr valign="top"><td><code>visu</code></td>
<td>
if <code>visu=TRUE</code> plot of number of empty clusters
is produced.</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
additional arguments for underlying <code>mfuzz</code></td></tr>
</table>

<h3>Details</h3>

<p>
A soft cluster is considered as empty, if none of the genes
has a correponding membership value larger than 0.5
</p>


<h3>Value</h3>

<p>
The function generates the matrix of the number of empty clusters
detected.</p>

<h3>Author(s)</h3>

<p>
Matthias E. Futschik (<a href="http://itb.biologie.hu-berlin.de/~futschik">http://itb.biologie.hu-berlin.de/~futschik</a>)
</p>


<h3>References</h3>

<p>
M.E. Futschik and B. Charlisle, Noise robust clustering of
gene expression time-course data, Journal of Bioinformatics and
Computational Biology, 3 (4), 2005
</p>


<h3>Examples</h3>

<pre>
data(yeast)
# Data pre-processing
yeastF &lt;- filter.NA(yeast)
yeastF &lt;- fill.NA(yeastF)
yeastF &lt;- standardise(yeastF)

#### parameter selection
yeastFR &lt;- randomise(yeastF)
cl &lt;- mfuzz(yeastFR,c=16,m=1.25)
mfuzz.plot(yeastFR,cl=cl,mfrow=c(4,5))

tmp  &lt;- partcoef(yeastFR)
F &lt;- tmp[[1]];F.n &lt;- tmp[[2]];F.min &lt;- tmp[[3]]
F &gt; F.min
</pre>



<hr><div align="center">[Package <em>Mfuzz</em> version 0.0.2 <a href="00Index.html">Index]</a></div>

</body></html>
